File: src/jastudio_src/jastudio/note/sentences/caching_sentence_configuration_field.py
from __future__ import annotations

from typing import TYPE_CHECKING

from autoslot import Slots  # pyright: ignore[reportMissingTypeStubs]
from typed_linq_collections.collections.q_set import QSet

from jastudio.ankiutils import app
from jastudio.note.note_constants import SentenceNoteFields
from jastudio.note.notefields.mutable_string_field import MutableStringField
from jastudio.note.sentences.sentence_configuration import SentenceConfiguration
from jastudio.sysutils.weak_ref import WeakRef, WeakRefable

if TYPE_CHECKING:
    from typed_linq_collections.collections.q_unique_list import QUniqueList

    from jastudio.note.sentences.sentencenote import SentenceNote
    from jastudio.note.sentences.word_exclusion_set import WordExclusionSet
    from jastudio.note.vocabulary.vocabnote import VocabNote

class CachingSentenceConfigurationField(WeakRefable, Slots):
    def __init__(self, sentence: WeakRef[SentenceNote]) -> None:
        self._sentence: WeakRef[SentenceNote] = sentence
        self.field: MutableStringField = MutableStringField(sentence, SentenceNoteFields.configuration)
        self._value: SentenceConfiguration = SentenceConfiguration.serializer().deserialize(self.field.value, self._save)

    @property
    def configuration(self) -> SentenceConfiguration:
        return self._value

    @property
    def incorrect_matches(self) -> WordExclusionSet: return self._value.incorrect_matches

    @property
    def hidden_matches(self) -> WordExclusionSet: return self._value.hidden_matches

    def highlighted_words(self) -> QUniqueList[str]: return self._value.highlighted_words

    @property
    def highlighted_vocab(self) -> QSet[VocabNote]:
        return QSet(vocab for vocab in (app.col().vocab.with_any_form_in(list(self.highlighted_words())))
                    if vocab.get_id() in self._sentence().parsing_result.get().matched_vocab_ids)

    def remove_highlighted_word(self, word: str) -> None:
        self.highlighted_words().discard(word)
        self._save()

    def reset_highlighted_words(self) -> None:
        self._value.highlighted_words.clear()
        self._save()

    def add_highlighted_word(self, vocab: str) -> None:
        self.highlighted_words().add(vocab.strip())
        self._save()

    def _save(self) -> None:
        self.field.set(SentenceConfiguration.serializer().serialize(self._value))
        self._sentence().update_parsed_words(force=True)

File: src/src_dotnet/JAStudio.Core/Note/Sentences/CachingSentenceConfigurationField.cs
using System;
using System.Collections.Generic;
using System.Linq;
using JAStudio.Core.Note.CorpusData;
using JAStudio.Core.Note.Vocabulary;

namespace JAStudio.Core.Note.Sentences;

public class CachingSentenceConfigurationField
{
   readonly SentenceNote _sentence;
   readonly NoteGuard _guard;

   public CachingSentenceConfigurationField(SentenceNote sentence, SentenceConfigSubData? data, NoteGuard guard)
   {
      _sentence = sentence;
      _guard = guard;
      Configuration = SentenceData.CreateConfiguration(data, () => _guard.Update(() => _sentence.UpdateParsedWords(force: true)));
   }

   public SentenceConfiguration Configuration { get; private set; }

   public WordExclusionSet IncorrectMatches => Configuration.IncorrectMatches;

   public WordExclusionSet HiddenMatches => Configuration.HiddenMatches;

   public List<string> HighlightedWords => Configuration.HighlightedWords;

   public HashSet<VocabNote> HighlightedVocab()
   {
      var highlightedWordsList = HighlightedWords.ToList();
      var vocabWithForms = _sentence.Services.Collection.Vocab.WithAnyFormIn(highlightedWordsList);
      var matchedVocabIds = _sentence.GetParsingResult().MatchedVocabIds;

      return vocabWithForms
            .Where(vocab => matchedVocabIds.Contains(vocab.GetId()))
            .ToHashSet();
   }

   public void RemoveHighlightedWord(string word) => _guard.Update(() =>
   {
      HighlightedWords.Remove(word);
      _sentence.UpdateParsedWords(force: true);
   });

   public void ResetHighlightedWords() => _guard.Update(() =>
   {
      Configuration.HighlightedWords.Clear();
      _sentence.UpdateParsedWords(force: true);
   });

   public void AddHighlightedWord(string vocab) => _guard.Update(() =>
   {
      HighlightedWords.Add(vocab.Trim());
      _sentence.UpdateParsedWords(force: true);
   });

   [Obsolete("For testing only")]
   public void SetValueDirectlyTestsOnly(SentenceConfiguration configuration)
   {
      Configuration = configuration;
   }
}
